### ISSUE TITLE
Use existing ComicInfo.XML when reading files for library import

### ISSUE THREAD

**Issue by MARAllMighty** [Casvt/Kapowarr#139]
As I'm sure others have as well, I have a comic book library that has already been scrapped against ComicVine and contains the appropriate .xml with all needed metadata. Would it be possible for the library importer to read this to help with parsing series info instead of just by file name or folder?

This would help me immensely with importing my existing library and allowing Kapowarr to download new issues on top of existing series, and would also help differentiate between series/oneshot/TPB/etc during import

**Comment by admlko** [comment]
I'd like to add this as a resource for implementing the feature: https://anansi-project.github.io/docs/comicinfo/intro

**Comment by Casvt** [comment]
@Gandalf-the-Blue that's basically issue #50.

**Comment by dylan-k** [comment]
I'd like to add that this option could also be helpful for cutting down on the number of comicvine API requests. If I understand it correctly, right now when you import a library it queries comicvine for every volume/series/folder. When there's a large library, that's already been scanned, those API requests could be skipped.

I don't know if Kapowarr can already check for an existing comicvineID in the metadata or filenames, but this could also help to make more efficient use of the API.

**Comment by Casvt** [comment]
@dylan-k eh it's not that more efficient than you might think.

Reading the file has two advantages:

1. Better data to search with. Instead of extracting and converting the data from the filename, we can use the properly formatted data from the metadata file.
2. No need to search at all if the CV ID is supplied in the metadata file. This indeed saves on an API call.

However, bullet 1 does not reduce the API requests at all and bullet 2 saves at most 1 API call per _series_. What limits the Library Import (in regard to the API limit) is adding the volumes to the library, not the searching.

**Comment by DuvelCorp** [comment]
Have a large collection of BDs (Bandes Dessinées).

Attempted to load my library into kapowarr, and it has been a huge failure. It's just 0 match at each attempt. While all my collection is properly handled with Calibre + Comicvine pluggin, and all files are CBZ, and all CBZ contain a proper ComicInfo.xml.

Can't understand how this is not handled natively as a comicinfo.xml is just the #1 rule-of-thumb and accepted standard in term of comic management....
I will thus come later I guess

**Comment by Casvt** [comment]
0 matches on everything? Have you not just reached the rate limit of ComicVine? Try again in an hour and if it starts matching again it's just the rate limit. If it doesn't, then share an example filepath I'll see why it doesn't work. 

Btw, most media downloaded from GetComics already has a metadata file inside the archive, so a lot of media will work just fine in reader softwares.

**Comment by DuvelCorp** [comment]
Nope I didn't reach it the comicvine API limits. I worked in several batches as suggested by default Kapowarr settings
Either there were no match at all, or the match was incorrect. And this is perfectly normal following my naming scheme if you only rely on this.... there's no editor/author/drawer/year in file names for instance, all those are the comicvine.xml.

You assumption of "all media from Getcomics has a metadata file inside the archive" is dangerous.  First its probably not always the case (and I must say I dont really use Getcomics so i dont know). Then You assume that all media will be downloaded from Getcomics, however I have seen from your roadmap that prowlarr/jacket indexers and torrent/usenet downloads are on the 'soon" roadmap (which is normal for a "arr" application), but then how do you think it will end up with those?

Maybe I dont understand what Kapowarr is or aim to be at all.  But if you want to be the radarr/sonarr/lidarr/readarr for comics, it means that it becomes the angular stone of a user's comics library management, replacing comicrack/mylar/calibre, and that proper metadata handling is thus an absolute necessity and can not be disregarded in anyway.

Radarr/Sonarr generates NFO from TMDB/TVDB
Liadarr read and embed metadata from Musicbrainz in audio files
Kapowarr should thus handle both import and export of comicinfo.xml

**Comment by Casvt** [comment]
> And this is perfectly normal following my naming scheme if you only rely on this.... there's no editor/author/drawer/year in file names for instance, all those are the comicvine.xml.

Kapowarr doesn't match based on editor, author or drawer. It uses the series title, year and volume number. As I asked for before, please supply an example filepath and I'll check why no match for it is found because it works fine for a lot of users.

---------------

And I'm sorry but I cannot resist not addressing the way you communicate. You come here generally talking aggressively and using an impressive 550 words disparaging me and my software while having no intention to solve the problem at hand. All this text and all you said was "I would really really like this feature" with a whole lot of insults in between. I work on this project in my free time, for free, with zero contributors. If you want it so bad, then either contribute, pay me or wait. And a lot of your criticism is based on incorrect assumptions or straight up wrong facts. If you're going to insult me, at least do it with facts. A compilation:

> I am not quite sure how you can even think that its possible to properly identify a comic with only the folder or file name, but in case you have a doubt: its impossible to have 100% match unless the Comicvine ID or ISBN is in the filename.

It's just straight up wrong to say it's 100% impossible. Simple counterexample: if the file says it's for Batman 2016, then Kapowarr will have a guaranteed correct match because there's only one Batman volume from 2016. Entire libraries of users are imported correctly and just fine because with a series name and volume number or year you're pretty much there. Especially if the series name is a bit unique. Most matching problems are in scenarios such as TPB's that have the same name, year and volume number as the original volume. But even then Kapowarr has some extra checks to often correctly figure out the right match.

> So I can't understand how comicinfo.xml are not handled natively (both for import and export btw) as it is just the rule-of-thumb and accepted standard in terms of comic metadata management for the last decade.

It's not like it's never happening. It's planned. Metadata support is important, but so is support for usenet and torrent indexers, proper ownership and permissions and a stable software. Metadata support is literally the next planned big feature after usenet and torrents. And what does the age of the standard have to do with how quickly it has to be integrated? In that case I'll have to hurry working on adding support for the ed2k protocol, because it's over 25 years old; a lot more than 10...

> Either there were no match at all, or the match was incorrect. And this is perfectly normal following my naming scheme if you only rely on this....

How can you know? Did you run the filenames through the extraction algorithm to see how it has processed them and then the searching algorithm to see how it matches them to volumes? You've made the blind assumption that your files are unmatchable to their volumes based on filename, and based on that blind assumption you start ranting about why support for metadata files is needed. This is why I asked for an example filepath, so that I can check for myself and maybe with a simple fix make it all work.

> You assumption of "all media from Getcomics has a metadata file inside the archive" is dangerous.

This is straight up wrong. You put quotes around text, as if I said it, that I in fact didn't write. I said that "**_MOST_**" media from GetComics has a metadata file. I get that not all has it, but a lot does. You disparage me on something I didn't say.

> First its probably not always the case (and I must say I dont really use Getcomics so i dont know).

The irony. You say you don't use GetComics and don't know, but in the very same sentence say that it's "probably not always the case". How can you know that if you say in the same sentence that you don't know?

> Then You assume that all media will be downloaded from Getcomics

Straight up wrong. I never said that. I said: "[...] most media downloaded from GetComics already has a metadata file inside the archive, so a lot of media will work just fine in reader softwares." Where in this sentence did I say that I'm assuming that all media is downloaded via GetComics? I said that most media from GetComics a metadata file has, so most media from GetComics will work in reader software. GetComics is currently the only download source that Kapowarr has, so most media downloaded by Kapowarr will have a metadata file. But I recognise that not _all_ media from GetComics comes with a metadata file, a lot of people have an existing library, and may still download media using external sources. 

> [...] however I have seen from your roadmap that prowlarr/jacket indexers and torrent/usenet downloads are on the 'soon" roadmap (which is normal for a "arr" application), but then how do you think it will end up with those?

Well firstly, quite a few torrent and usenet downloads also include metadata files. But I realise that it will be a lot less than with GetComics. And that is why after adding support for the mentioned downloading methods, the very next big feature is metadata support. You could've seen that on the project board that you love to look at.

> But if you want to be the radarr/sonarr/lidarr/readarr for comics, it means that it becomes the angular stone of a user's comics library management, replacing comicrack/mylar/calibre, and that proper metadata handling is thus an absolute necessity and can not be disregarded in anyway.

That is indeed what I want it to become. And when I have time available, I work on this project to come closer to this goal. Your 550 words have already made clear that you find it an absolute necessity. And who said I'm disregarding this? It's literally going to be added. You're acting like it's never going to be added or something. 

---------------

So next time you're insulting someone, whether it be me or someone else, at least do it properly and fairly. In case you forgot after reading all this text: if you share an example filepath, I can check it out and maybe fix it.

**Comment by wearysky** [comment]
> Btw, most media downloaded from GetComics already has a metadata file inside the archive, so a lot of media will work just fine in reader softwares.

This is true, but that doesn't necessarily mean it will be accurate or complete. Just to give an example, say if somebody downloaded Uncanny X-Men (2024 - vol 6 according to ComicVine) issue 10 from there recently... It does indeed contain a ComicInfo.xml file! Unfortunately, that file doesn't contain volume information. Issue 11 does include a volume in its ComicInfo.xml but it's wrong (it's parsed out of the ComicVine URL for this volume, instead of being the actual volume). And all the rest of the issues for that volume downloaded from GetComics have the volume listed as "2024" instead of Volume 6, so they're wrong too but at least they group together in comics readers, whereas 10 and 11 end up somewhere else entirely. 

All that is to say, it would be nice to eventually be able to write out files for a given volume, say, so that it's consistent for all issues. Eventually, when you get around to looking at this that is :)

**Comment by trumblejoe** [comment]
To add to this thread, reading existing metadata is great, writing metadata would be a killer feature as only Mylarr and Comictagger and a few other niche apps are currently able too- but none are easy (except ComicTagger but that's a desktop app making server use difficult). An additional ability to create non-existing series matches on import would be an almost end-to-end capability. Download it, tag it, create series if needed, auto or manually match if series found. All of this comes back to being able to use the file metadata on import and would propel the app forward while being able to hijack into mostly existing functions.
